backend/error_source_middleware.go

```diff
@@ -0,0 +1,132 @@
+package backend
+
+import (
+	"context"
+	"errors"
+	"fmt"
+)
+
+// NewErrorSourceMiddleware returns a new backend.HandlerMiddleware that sets the error source in the
+// context.Context, based on returned errors or query data response errors.
+// If at least one query data response has a "downstream" error source and there isn't one with a "plugin" error source,
+// the error source in the context is set to "downstream".
+func NewErrorSourceMiddleware() HandlerMiddleware {
+	return HandlerMiddlewareFunc(func(next Handler) Handler {
+		return &ErrorSourceMiddleware{
+			BaseHandler: NewBaseHandler(next),
+		}
+	})
+}
+
+type ErrorSourceMiddleware struct {
+	BaseHandler
+}
+
+func (m *ErrorSourceMiddleware) handleDownstreamError(ctx context.Context, err error) error {
+	if err == nil {
+		return nil
+	}
+
+	if IsDownstreamError(err) {
+		if innerErr := WithDownstreamErrorSource(ctx); innerErr != nil {
+			return fmt.Errorf("failed to set downstream error source: %w", errors.Join(innerErr, err))
+		}
+	}
+
+	return err
+}
+
+func (m *ErrorSourceMiddleware) QueryData(ctx context.Context, req *QueryDataRequest) (*QueryDataResponse, error) {
+	resp, err := m.BaseHandler.QueryData(ctx, req)
+	err = m.handleDownstreamError(ctx, err)
+
+	if err != nil || resp == nil || len(resp.Responses) == 0 {
+		return resp, err
+	}
+
+	// Set downstream error source in the context if there's at least one response with downstream error source,
+	// and if there's no plugin error
+	var hasPluginError bool
+	var hasDownstreamError bool
+	for refID, r := range resp.Responses {
+		if r.Error == nil {
+			continue
+		}
+
+		if !r.ErrorSource.IsValid() {
+			// if the error is a downstream error, set error source to downstream, otherwise plugin.
+			if IsDownstreamError(r.Error) {
+				r.ErrorSource = ErrorSourceDownstream
+			} else {
+				r.ErrorSource = ErrorSourcePlugin
+			}
+			resp.Responses[refID] = r
+		}
+
+		if !r.Status.IsValid() {
+			r.Status = statusFromError(r.Error)
+			resp.Responses[refID] = r
+		}
+
+		if r.ErrorSource == ErrorSourceDownstream {
+			hasDownstreamError = true
+		} else {
+			hasPluginError = true
+		}
+	}
+
+	// A plugin error has higher priority than a downstream error,
+	// so set to downstream only if there's no plugin error
+	if hasDownstreamError && !hasPluginError {
+		if err := WithDownstreamErrorSource(ctx); err != nil {
+			return resp, fmt.Errorf("failed to set downstream status source: %w", err)
+		}
+	}
+
```

Comment:
Just to be clear, we don't care about setting a plugin error source here, right? This will be the default anyway if not set? 🤔

Reply:
Yes, exactly. default init here https://github.com/grafana/grafana-plugin-sdk-go/blob/64eea532a8b15269e85a6d507dc8f9f8e9909a4f/backend/handler_middleware.go#L50